packages/react/src/file-upload/FileUpload.tsx

```diff
@@ -102,28 +105,95 @@ export const FileUpload = forwardRef<HTMLInputElement, FileUploadProps>(
 		const maxSizeBytes = maxSize && !isNaN(maxSize) ? maxSize * 1000 : 0;
 		const formattedMaxFileSize = formatFileSize(maxSizeBytes);
 
-		const [fileRejections, setFileRejections] = useState<RejectedFile[]>([]);
+		if (maxFiles !== undefined && maxFiles < 1) {
+			console.warn(
+				'FileUpload: maxFiles cannot be less than 1. The property is being ignored.'
+			);
+		}
 
-		const handleRemoveFile = (file: FileWithStatus) => {
-			const indexOfFile = value.indexOf(file);
-			onChange(value.filter((_, index) => index !== indexOfFile));
-		};
+		const [acceptedFiles, setAcceptedFiles] = useState(value);
+		const [tooManyFilesRejections, setTooManyFilesRejections] = useState<
+			FileRejection[]
+		>([]);
+		const [invalidRejections, setInvalidRejections] = useState<FileRejection[]>(
+			[]
+		);
 
-		const handleRemoveRejection = (fileName: string) => {
-			setFileRejections(
-				fileRejections.filter((err) => err.file.name !== fileName)
-			);
+		const allRejections = [...invalidRejections, ...tooManyFilesRejections];
+
+		const handleRemoveAcceptedFile = (index: number) => {
+			clearErrors();
+
+			setAcceptedFiles((prevAcceptedFiles) => {
+				const updatedAcceptedFiles = removeItemAtIndex(
+					prevAcceptedFiles,
+					index
+				);
+				return updatedAcceptedFiles;
+			});
+
+			onChange?.(acceptedFiles);
 		};
 
 		const handleDropAccepted = (acceptedFiles: FileWithStatus[]) => {
-			// replace file if multiple is false
+			clearErrors();
+
 			if (multiple) {
-				onChange([...value, ...acceptedFiles]);
+				setAcceptedFiles((prevAcceptedFiles) => {
+					const prevAcceptedFilesSet = new Set(
+						prevAcceptedFiles.map(
+							(file) => `${file.name}-${file.size}-${file.type}`
+						)
+					);
+					const filteredAcceptedFiles = acceptedFiles.filter(
+						(file) =>
+							!prevAcceptedFilesSet.has(
+								`${file.name}-${file.size}-${file.type}`
+							)
+					);
+
+					const newAcceptedFiles = [
+						...prevAcceptedFiles,
+						...filteredAcceptedFiles,
+					];
+
+					if (maxFiles && newAcceptedFiles.length > maxFiles) {
+						setTooManyFilesRejections((prevTooManyFilesList) => {
+							const newTooManyFilesList = newAcceptedFiles
+								.slice(maxFiles)
+								.map(convertFileToTooManyFilesRejection);
+
+							if (
+								JSON.stringify(prevTooManyFilesList) ===
+								JSON.stringify(newTooManyFilesList)
+							) {
+								return prevTooManyFilesList;
+							}
+
+							return [
+								...prevTooManyFilesList,
+								...newAcceptedFiles
+									.slice(maxFiles)
+									.map(convertFileToTooManyFilesRejection),
+							];
+						});
+						return newAcceptedFiles.slice(0, maxFiles);
+					}
+
+					return newAcceptedFiles;
```

Comment:
@ChrisLaneAU My brain cannot parse this. I don't know if parts of this (like the `prevTooManyFilesList` stuff) are necessary or over engineered Can you please check it?